Replace safetyNey to Play Integrity (#2371)
OWASP/owasp-mastg: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md

- MEETS_VIRTUAL_INTEGRITY: Running on Emulator with Google Play Services: In this scenario, the app is running in a simulated environment of an Android device that has Google Play Services. The emulator was checked and approved for system integrity and meets the necessary Android compatibility requirements.

**API Errors:**

The API can return local errors such as "APP_NOT_INSTALLED" and "APP_UID_MISMATCH", which can indicate a fraud attempt or attack. In addition, outdated Google Play Services or Play Store can also cause errors, and it is important to check these situations to ensure proper integrity verification functionality and to ensure the environment is not intentionally set up for an attack. You can find more details on the [official page](https://developer.android.com/google/play/integrity/error-codes).

**Best practices:**

1. Although Play Integrity provides an additional layer of security for Android applications, it is not a single security solution. It is important to have complementary measures in place, such as input data validation, user authentication and anti-fraud protection.

> 2. It is important to avoid requesting attests too frequently in order to minimize the impact on device resources. Furthermore, it is good practice to use the API only in situations where device integrity verification is truly necessary.

Reviewer: ```suggestion
2. It is important to avoid querying the Play Protect API too frequently in order to minimize the impact on device resources. Furthermore, it is good practice to use the API only in situations where device integrity verification is truly necessary.

```

Author: I fully agree.

> 3. By including a "NONCE" in the request, the verification server can ensure that the response matches the original request and has not been tampered with by third parties. It is a random value generated by the app (or server) and sent with the attest request.

Reviewer: Same comment here for usage of 'attest'.

> 3. By including a "NONCE" in the request, the verification server can ensure that the response matches the original request and has not been tampered with by third parties. It is a random value generated by the app (or server) and sent with the attest request.

Reviewer: ```suggestion
3. By including a "NONCE" in the request, the verification server can ensure that the response matches the original request and has not been tampered with by third parties. It is a random value generated by the app (or server) and sent with the integrity verification request.

```

Root detection can also be implemented through libraries such as [RootBeer](https://github.com/scottyab/rootbeer "RootBeer").

#### Google Play Integrity
  
Google has launched the Google Play Integrity API to improve the security and integrity of apps and games on Android starting from Android 4.4 (level 19). The previous official API, [SafetyNet](https://developer.android.com/training/safetynet), did not cover all the security needs that Google wanted for the platform, so Play Integrity was developed with the basic functions of the previous API and integrated additional features. This change aims to protect users against dangerous and fraudulent interactions.

**Google Play Integrity provides the following protections:**

Verification of genuine Android device: verifies that the app is running on a legitimate Android device.  
User license validation: Indicates whether the user installed or paid for the app or game via the Google Play Store.  
Unmodified binary verification: determines if the app is interacting with the original binary recognized by Google Play.

> 

Reviewer 2: I think this could be reworded as:

Google Play Integrity offers the following safeguards:

- **Verification of genuine Android device**: It verifies that the application is running on a legitimate Android device.
- **User license validation**: It indicates whether the user installed or purchased the application or game through the Google Play Store
- **Unmodified binary verification**: It determines whether the application is interacting with the original binary recognized by Google Play.

Author: Fixed as requested

The API provides four macro categories of information to help the security team make a decision. These categories include:

> 1.Request Details: In this section, details are obtained about the app package that requested the integrity check, including  its format (my.package.com), a base64-encoded ID provided by the developer to establish a link between the request and the integrity certificate, and the time of the request execution in milliseconds.

Reviewer 2: Consider putting **Request Details** in bold, and apply the same style with points `2`, `3`, and `4`.

Reviewer 3: Also please add a space after the dot, like this `1. `. If not, it won't be recognized as a list item.

4.Device Integrity: This section presents information that verifies the authenticity of the Android environment in which the app is running.

- MEETS_DEVICE_INTEGRITY: Running on Android Device with Google Play Services: The app is running on an Android device with Google Play Services. This device passed system integrity checks and meets the Android compatibility requirements.
- MEETS_BASIC_INTEGRITY: Running on Device without Google Play Services: The app is running on a device that passed basic system integrity checks. This may occur because the device may be running an unrecognized version of Android, has an unlocked bootloader, or has not been certified by the manufacturer.
- MEETS_STRONG_INTEGRITY: The app is running on an Android device equipped with Google Play Services technology, ensuring a strong system integrity such as a hardware-protected boot process.

> - MEETS_VIRTUAL_INTEGRITY: Running on Emulator with Google Play Services: In this scenario, the app is running in a simulated environment of an Android device that has Google Play Services. The emulator was checked and approved for system integrity and meets the necessary Android compatibility requirements.

Reviewer 2: Consider putting `MEETS_DEVICE_INTEGRITY`, `MEETS_BASIC_INTEGRITY`, `MEETS_STRONG_INTEGRITY` and `MEETS_VIRTUAL_INTEGRITY` in `code`

Author: Fixed as requested

**API Errors:**

> The API can return local errors such as "APP_NOT_INSTALLED" and "APP_UID_MISMATCH", which can indicate a fraud attempt or attack. In addition, outdated Google Play Services or Play Store can also cause errors, and it is important to check these situations to ensure proper integrity verification functionality and to ensure the environment is not intentionally set up for an attack. You can find more details on the [official page](https://developer.android.com/google/play/integrity/error-codes).

Reviewer 2: Suggest to use `APP_NOT_INSTALLED` and `APP_UID_MISMATCH` in `code` rather than using quotes

**Best practices:**

1. Although Play Integrity provides an additional layer of security for Android applications, it is not a single security solution. It is important to have complementary measures in place, such as input data validation, user authentication and anti-fraud protection.
2. It is important to avoid querying the Play Protect API too frequently in order to minimize the impact on device resources. Furthermore, it is good practice to use the API only in situations where device integrity verification is truly necessary.

> 3. By including a "NONCE" in the request, the verification server can ensure that the response matches the original request and has not been tampered with by third parties. It is a random value generated by the app (or server) and sent with the integrity verification request.

Reviewer 2: Suggest to use `NONCE` rather than "NONCE"

Author: Fixed as requested

**Best practices:**

1. Although Play Integrity provides an additional layer of security for Android applications, it is not a single security solution. It is important to have complementary measures in place, such as input data validation, user authentication and anti-fraud protection.
2. It is important to avoid querying the Play Protect API too frequently in order to minimize the impact on device resources. Furthermore, it is good practice to use the API only in situations where device integrity verification is truly necessary.

3. By including a `NONCE` in the request, the verification server can ensure that the response matches the original request and has not been tampered with by third parties. It is a random value generated by the app (or server) and sent with the integrity verification request.

**Limitations:**  
It is important to note that there is a daily limit for the Google Play Services Integrity Verification API requests. By default, up to 10,000 requests are allowed per day, but for applications that require a higher number of requests, it is necessary to contact Google to request an expansion of this limit.

> **Example:**  

Reviewer 3: ```suggestion
**Example Request:**  
```

- MSTG-RESILIENCE-9: "Obfuscation is applied to programmatic defenses, which in turn impede de-obfuscation via dynamic analysis."
- MSTG-RESILIENCE-10: "The app implements a 'device binding' functionality using a device fingerprint derived from multiple properties unique to the device."

### Play Integrity Attestation

- Developer Guideline - <https://developer.android.com/google/play/integrity/overview>
- Play Integrity API - Verdict - <https://developer.android.com/google/play/integrity/verdict>

> - Play Integrity API - Quota Request - <https://developer.android.com/google/play/integrity/overview?#usage-tiers>

Reviewer 2: Should we use `[Text ](URL "Name")` format rather than this one?
